graphql_compiler/interpreter/debugging.py

```diff
@@ -0,0 +1,196 @@
+    the start parameter. In this function, we skip elements with indices in [0, first_index)
+    in the sequence, and only iterate over elements with indices in [first_index, len(sequence)).
+
+    Args:
+        sequence: arbitrary sequence of values over which to iterate
+        first_index: the first index of the sequence to produce in the resulting iterable
+
+    Yields:
+        tuples (index, element_at_that_index_in_sequence), analogous to the built-in enumerate()
+    """
+    for index in range(first_index, len(sequence)):
+        yield index, sequence[index]
+
+
+def _unzip_and_yield_second(iterable: Iterable[Tuple[T, U]]) -> Iterable[U]:
+    """Unpack an iterable of tuples and only yield the second element of each tuple."""
+    for _, second in iterable:
+        yield second
+
+
+@dataclass(frozen=True)
+class AdapterOperation:
+    """The record of an action performed to or by an InterpreterAdapter function as part of a trace.
+
+    For example, calling get_tokens_of_type("Foo") would be an AdapterOperation with kind "call",
+    with name "get_tokens_of_type", with unique ID and parent unique IDs appropriate for the trace,
+    and with data (("Foo",), {}) as the tuple of the positional arguments tuple and kwargs dict.
+
+    The parent_uid field allows us to track dependencies across different operations. For example,
+    data yielded by a call to a generator-style function will be recorded in AdapterOperations
+    with kind "yield" whose parent_uid points to the AdapterOperation of the particular invocation
+    of the function that is yielding that data.
+    """
+
+    kind: Literal["call", "yield", "return"]
+    name: str
+    uid: int
+    parent_uid: int
+    data: Any
+
+
+@dataclass(frozen=True)
+class RecordedTrace(Generic[DataToken]):
+    """A complete immutable recording of the execution of a query via an InterpreterAdapter.
+
+    Includes a linearized sequence of all operations performed during the execution of the query:
+    - adapter function calls, yields, and returns;
+    - actuation of all input and output iterables.
+
+    Traces may record complete executions of queries (i.e. ones where the generators are run
+    to completion), and may also be used to record partial executions (e.g. "fetch only 3 results").
+
+    The recording is sufficiently detailed to be used in at least the following use cases:
+    - to examine, either manually or through automated tests, a query's execution flow;
+    - to "impersonate" an InterpreterAdapter implementation for a given query, allowing bugs in
+      the interpreter library code to be isolated and reproduced without an InterpreterAdapter;
+    - to "impersonate" the interpreter library code toward an InterpreterAdapter implementation,
+      allowing adapter implementers to test and examine their implementations in isolation from
+      the rest of the interpreter library code.
```

Comment:
```suggestion
    The recording is sufficiently detailed to be used in at least the following use cases:
    - to examine, either manually or through automated tests, a query's execution flow;
    - to "impersonate" an InterpreterAdapter implementation for a given query, allowing bugs in
      the interpreter code to be isolated and reproduced without an InterpreterAdapter;
    - to "impersonate" the interpreter code toward an InterpreterAdapter implementation,
      allowing adapter implementers to test and examine their implementations in isolation from
      the rest of the interpreter code.
```
Since we've settled on the interpreter/adapter dichotomy, I think it's worth dropping "library" for brevity & clarity.  I think "library" connotes "something that I call" to a lot of developers, and its use might confuse people who are responsible for writing adapters but who may not be responsible for running the service in which their adapters are used.